extensions/sftp/processors/PutSFTP.cpp

```diff
@@ -0,0 +1,846 @@
+  std::stringstream target_path_ss;
+  target_path_ss << remote_path << "/";
+  if (!IsNullOrEmpty(temp_file_name)) {
+    target_path_ss << temp_file_name;
+  } else if (dot_rename_) {
+    target_path_ss << "." << resolved_filename;
+  } else {
+    target_path_ss << resolved_filename;
+  }
+  auto target_path = target_path_ss.str();
+  std::stringstream final_target_path_ss;
+  final_target_path_ss << remote_path << "/" << resolved_filename;
+  auto final_target_path = final_target_path_ss.str();
+  logger_->log_debug("The target path is %s, final target path is %s", target_path.c_str(), final_target_path.c_str());
+
+  ReadCallback read_callback(target_path.c_str(), client, conflict_resolution_);
+  session->read(flow_file, &read_callback);
+
+  if (!read_callback.commit()) {
+    session->transfer(flow_file, Failure);
+    return true;
+  }
+
+  /* Move file to its final place */
+  if (target_path != final_target_path) {
+    if (!client->rename(target_path, final_target_path, conflict_resolution_ == CONFLICT_RESOLUTION_REPLACE /*overwrite*/)) {
+      logger_->log_error("Failed to move temporary file %s to final path %s", target_path, final_target_path);
+      if (!client->removeFile(target_path)) {
+        logger_->log_error("Failed to remove temporary file %s", target_path.c_str());
+      }
+      session->transfer(flow_file, Failure);
+      return true;
+    }
+  }
+
+  /* Set file attributes if needed */
+  if (last_modified_time_set ||
+      permissions_set ||
+      remote_owner_set ||
+      remote_group_set) {
+    utils::SFTPClient::SFTPAttributes attrs;
+    attrs.flags = 0U;
+    if (last_modified_time_set) {
+      /*
+       * NiFi doesn't set atime, only mtime, but because they can only be set together,
+       * if we don't want to modify atime, we first have to get it.
+       * Therefore setting them both saves an extra protocol round.
+       */
+      attrs.flags |= utils::SFTPClient::SFTP_ATTRIBUTE_MTIME | utils::SFTPClient::SFTP_ATTRIBUTE_ATIME;
+      attrs.mtime = last_modified_time;
+      attrs.atime = last_modified_time;
+    }
+    if (permissions_set) {
+      attrs.flags |= utils::SFTPClient::SFTP_ATTRIBUTE_PERMISSIONS;
+      attrs.permissions = permissions;
+    }
+    if (remote_owner_set) {
+      attrs.flags |= utils::SFTPClient::SFTP_ATTRIBUTE_UID;
+      attrs.uid = remote_owner;
+    }
+    if (remote_group_set) {
+      attrs.flags |= utils::SFTPClient::SFTP_ATTRIBUTE_GID;
+      attrs.gid = remote_group;
+    }
+    if (!client->setAttributes(final_target_path, attrs)) {
+      /* This is not fatal, just log a warning */
+      logger_->log_warn("Failed to set file attributes for %s", target_path);
+    }
+  }
+
+  session->transfer(flow_file, Success);
+  put_connection_back_to_cache();
+  return true;
+}
+
+void PutSFTP::onTrigger(const std::shared_ptr<core::ProcessContext> &context, const std::shared_ptr<core::ProcessSession> &session) {
+  for (size_t i = 0; i < batch_size_; i++) {
```

Comment:
Please add an unsigned validator for BatchSize property. 
Moreover 0 should either be forbidden or we should process as many flowfiles as available. 
Doing the latter does have consequences, please check this PR, quite the same topic was discussed:
https://github.com/apache/nifi-minifi-cpp/pull/553

Reply:
Nice catch, thank you!

Reply:
0 is treated as unlimited (uint64 max), in accordance with the NiFi implementation.